Tradheli-improvement to internal rotor speed governor

bnsgeyer opened the pull request:
This PR provides a significant enhancement to the internal rotor speed governor.  It enables automatic spool up from ground idle to throttle unlimited.   it also provides cool down period for internal combustion engines.  Many thanks to Chris Olson (@MidwestAire) for his work on this and others who helped test it.

MidwestAire reviewed (commented):
@bnsgeyer I noted the param macro index conflict with the SITL autorotation param. Otherwise the code looks intact after the rebase. Since the autothrottle logic is quite complicated I think it would be a good idea for me to build this and run it in a real helicopter before the PR is merged.
Thanks!

MidwestAire commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -231,18 +241,39 @@ void AP_MotorsHeli_RSC::output(RotorControlState state)
+            _governor_output = 0.0f;
+            _governor_torque_reference = 0.0f;
+            if (!_autothrottle) {
+                if (_in_autorotaion) {
```
@bnsgeyer "autorotation" is spelled incorrectly here. This came from the autorotation code and was merged here during the rebase on master. Should make sure this if() statement executes correctly.

bnsgeyer replied:
The autorotation code only applies for SITL.   Additionally it only applies to the SETPOINT RSC mode because this particular if statement is placing the idle throttle output in the autorotation PWM that ESC's want to see so they spool up immediately rather than slow start up.  I'm pretty confident that I merged it correctly.  I do want you to test this branch to ensure the autothrottle functionality has not been altered by the merge with master.

MidwestAire replied:
Great will do that. I saw misspelling and thought maybe was accidental and would not execute properly. Have not looked at rest of auto code to see if misspelled thru out.  Hopefully can build and test tomorrow on master.

MidwestAire commented:
@bnsgeyer I ran this early this morning in one of my piston engine helicopters. As you are aware, the governor portion itself is not new, we have flown that for over a year. The main concern was the autothrottle sequencing, fault lockout, and integration with the spool logic in 4.1-dev. That functions perfectly after the rebase. So from a safety and functionality standpoint, this is good to go on my end for 4.1-dev. Works as I designed it to.

Gone4Dirt reviewed (commented):
I have offered a few comments.  I hope it helps.

MidwestAire commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -418,3 +424,70 @@ float AP_MotorsHeli_RSC::calculate_desired_throttle(float collective_in)
+            if (_governor_fault_count > 200.0f) {
+                _governor_fault = true;
+                _governor_engage = false;
+                _governor_output = 0.0f;
```
Matt, there is only two variables that could be reset in such a function. The flags can't be in it, because they depend on the status on the internal fault checking and spool logic. By the time you declare the function, write the function, and put in calls to it in four places it is more lines of code than just inline-resetting the two variables in four places. That's why I didn't do it.

bnsgeyer replied:
done.  Thanks!

MidwestAire commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -418,3 +424,70 @@ float AP_MotorsHeli_RSC::calculate_desired_throttle(float collective_in)
+            _governor_output = (_rotor_rpm / _governor_rpm) * torque_limit;
+            _control_output = constrain_float(throttlecurve + _governor_output, 0.0f, 1.0f);
+            _governor_torque_reference = _control_output;  // increment torque setting to be passed to main power loop
+            if (_rotor_rpm >= (_governor_rpm - 2.0f)) {
```
No, it can't be a percentage. Percentage builds in "fudge factor". On a rotor that is turning at 540 rpm +/- 0.5% is +/- 2.7 rpm. On a rotor turning at 2,000 rpm +/- 0.5% is +/- 10 rpm. While the percentage would work ok on the 540 rpm rotor, it would be huge "fudge factor" on the 2,000 rpm rotor.

The governor's torque reference is the equivalent of manifold pressure on a piston engine, torque setting on a turbine. It has to start moving it when the rotor rpm is +/-2 or it don't work. Been there, done that, tried it when the governor was designed over a year ago and it has flown slightly over 1,100 hours before it came to ArduPilot. The existing 4 rpm of hysteresis is to deal with the output of the rpm library not returning exactly the same value every time thru the loop. If it weren't for that it would be zero tolerance instead of +/-2

MidwestAire commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -125,61 +125,72 @@ const AP_Param::GroupInfo AP_MotorsHeli_RSC::var_info[] = {
-    // @Range: 0 40
+    // @Param: GOV_FF
+    // @DisplayName: Governor Feedforward
+    // @Description: Feedforward governor gain to throttle response during sudden loading/unloading of the rotor system. If Rrpm drops excessively during full collective climb with the droop response set correctly, increase the governor feedforward. If Rrpm drops excessively under heavy load also inspect the setting for maximum collective pitch to ensure that the pitch setting corresponds to maxi ...
```
Matt, it certainly can be in the wiki, and will be. However, I take note of what Paul has done in the EKF libraries in including the appropriate information in the metadata for users to understand how to make a setting, and what it does. When you are out in the field and tuning something, a good ground station like QGC displays this quite nicely on the right sidebar when you select the setting. Having that information at your fingertips is much more user-friendly that having to refer to a wiki where you may not have internet access. And nothing I hate more than a setting with metadata that says "this is for advanced users only". It has zero impact on cpu usage. Huge impact on user-friendliness, which many times ArduPilot lacks.

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -125,61 +125,72 @@ const AP_Param::GroupInfo AP_MotorsHeli_RSC::var_info[] = {
     // @Increment: 10
     // @User: Standard
-    AP_GROUPINFO("GOV_SETPNT", 13, AP_MotorsHeli_RSC, _governor_reference, AP_MOTORS_HELI_RSC_GOVERNOR_SETPNT_DEFAULT),
+    AP_GROUPINFO("GOV_RPM", 13, AP_MotorsHeli_RSC, _governor_rpm, AP_MOTORS_HELI_RSC_GOVERNOR_RPM_DEFAULT),
```
I think the _governor_reference is an AP_Int16 but the new _governor_rpm is an AP_Float so I think this should be moved to a new eeprom slot.

bnsgeyer replied:
done. thanks!

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli_RSC.h:
```diff
@@ -173,24 +175,29 @@ class AP_MotorsHeli_RSC {
-    float       get_governor_disengage() { return _governor_disengage * 0.01; }
-    float       get_governor_droop_response() { return _governor_droop_response * 0.01; }
-    float       get_governor_tcgain() { return _governor_tcgain * 0.01; }
+    float       get_idle_output() { return _idle_output * 0.01; }
```
these methods can all be const.

bnsgeyer replied:
done. thanks!

rmackay9 commented:
The changes are all tradheli specific so I have no real concerns about this going in.  Thanks to @Gone4Dirt for all the review comments as well.

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -125,61 +125,72 @@ const AP_Param::GroupInfo AP_MotorsHeli_RSC::var_info[] = {
     // @Increment: 10
     // @User: Standard
-    AP_GROUPINFO("GOV_SETPNT", 13, AP_MotorsHeli_RSC, _governor_reference, AP_MOTORS_HELI_RSC_GOVERNOR_SETPNT_DEFAULT),
+    AP_GROUPINFO("GOV_RPM", 13, AP_MotorsHeli_RSC, _governor_rpm, AP_MOTORS_HELI_RSC_GOVERNOR_RPM_DEFAULT),
```
Need to make this a new parameter rather than re-naming the existing parameter especially since it changed type (AP_INT16 to AP_Float)

MidwestAire commented:
@bnsgeyer I made some changes that addresses most of the comments here.
1) I removed two constrain_floats that were not needed
2) changed the governor's _control_output lower limit from a floating value to a hard limit
3) added the governor_reset() function as suggested by Matt. This increased the lines of code in the cpp by one line. But removing the governor defaults from the header and getting rid of unnecessary comment lines in RSC.h reduced the size of the header by 10 lines.
4) rearranged the params for the governor in alphabetical order by @Param name and assigned a new eeprom slot for the rpm. Reduced the verbosity of the param descriptions.

This set of changes is in my branch. If it looks good to you I'll push it to the PR branch
https://github.com/MidwestAire/ardupilot/commit/9cba692a7fb5fdb83336815c617b7ea54c51ec8a

Is it really necessary to have the defaults for the throttle curve as a #define in the header? Those could be moved to the AP_GROUPINFO in the cpp, as well, since they are never changed anyway.

Also, since this also intended as a general cleanup and overhaul of the RSC, function declarations are usually self-explanatory and re-commented in the C file where they are used. Could get rid of more unnecessary comment lines in the header and group other things, like the autorotation functions() and maybe the functions for rotor speeds under one comment line. Comments are fine, but when it starts getting too many of them, one for each declaration, it gets messy. I prefer how it was done at the bottom of the header in the public access specifier for the parameter assessors, it's a lot cleaner.

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -231,65 +248,55 @@ void AP_MotorsHeli_RSC::output(RotorControlState state)
             _control_output = 0.0f;
+
+            // governor is forced to disengage status and ensure governor outputs are reset
+            _autothrottle = false;
```
done.  Thanks!

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.h:
```diff
@@ -173,24 +175,29 @@ class AP_MotorsHeli_RSC {
+    AP_Float        _governor_droop_response;   // governor response to droop under load
+    AP_Float        _governor_ff;               // governor feedforward variable
+    AP_Float        _governor_fault_count;      // variable for tracking governor speed sensor faults
+    AP_Float        _autothrottle_fast_idle;    // autothrottle variable for cooldown
```
done. Thanks!

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.h:
```diff
@@ -173,24 +175,29 @@ class AP_MotorsHeli_RSC {
+    AP_Float        _governor_compensator;      // governor torque compensator variable
+    AP_Float        _governor_droop_response;   // governor response to droop under load
+    AP_Float        _governor_ff;               // governor feedforward variable
+    AP_Float        _governor_fault_count;      // variable for tracking governor speed sensor faults
```
done. Thanks!

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.h:
```diff
@@ -173,24 +175,29 @@ class AP_MotorsHeli_RSC {
+    AP_Float        _governor_ff;               // governor feedforward variable
+    AP_Float        _governor_fault_count;      // variable for tracking governor speed sensor faults
+    AP_Float        _autothrottle_fast_idle;    // autothrottle variable for cooldown
+    AP_Float        _autothrottle_cooldown;     // autothrottle cooldown timer to provide a fast idle
```
done. thanks!

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -212,6 +223,12 @@ void AP_MotorsHeli_RSC::set_throttle_curve()
 {
+    // _rotor_rpm available to the RSC output
+    const AP_RPM *rpm = AP_RPM::get_singleton();
+    if (!rpm->get_rpm(0, _rotor_rpm)) {
```
@Gone4Dirt would this be acceptable to check the nullptr or should I nest the rpm get statement in the nullptr check?

    if (rpm == nullptr || !rpm->get_rpm(0, _rotor_rpm)) {
        _rotor_rpm = -1;
    }

bnsgeyer replied:
Done!

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.h:
```diff
@@ -173,24 +175,29 @@ class AP_MotorsHeli_RSC {
-    AP_Float        _governor_tcgain;       // governor throttle curve weighting, range 50-100%
+    AP_Int16        _power_slewrate;            // throttle slew rate (percentage per second)
+    AP_Int16        _thrcrv[5];                 // throttle value sent to throttle servo at 0, 25, 50, 75 and 100 percent collective
+    AP_Float        _governor_rpm;              // governor reference for speed calculations
```
Changed to AP_Int16.  Thanks!

bnsgeyer commented:
@Gone4Dirt would like to get final approval from you for the changes I have implemented since your initial review.  Most of it was addressing your comments. I did add a fix for a bug I found in the spool up logic of the governor. I also moved the cool down timer out of the governor to make it generic to all RSC Modes.   Lastly I restored the governor range parameter. Chris had fixed this to -3/+2 %.  I have had some feedback that in some cases this is too restrictive and causes the governor to fault.

Gone4Dirt commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -125,61 +125,81 @@ const AP_Param::GroupInfo AP_MotorsHeli_RSC::var_info[] = {
     // @User: Standard
-    AP_GROUPINFO("GOV_DISGAG", 14, AP_MotorsHeli_RSC, _governor_disengage, AP_MOTORS_HELI_RSC_GOVERNOR_DISENGAGE_DEFAULT),
+    AP_GROUPINFO("AROT_PCT", 18, AP_MotorsHeli_RSC, _ext_gov_arot_pct, 0),
+#endif
```
FYI this is just Git messing up the Diff.  The AROT_PCT parameter is not changing from master to this PR.  Incase anyone else looking at this was a bit confused like I was...

bnsgeyer replied:
Yes it was strange how it showed the changes

Gone4Dirt reviewed (changes requested):
I am not experienced with governers so cannot review how it works.  Mostly trivial comments to tidy up.  I find the autothrottle_run extremly hard to follow due to the number of nested if{}, but I can see that it would be difficult to pull that appart.

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -231,65 +263,51 @@ void AP_MotorsHeli_RSC::output(RotorControlState state)
+                if (_cooldown_time > 0) {
+                    _control_output = get_idle_output() * 1.5f;
+                    _fast_idle_timer += dt;
+                    if (_fast_idle_timer > (float)_cooldown_time) {
```
I thought our desire is to keep params as Integers if at all possible to keep down eeprom size?

Gone4Dirt replied:
Yes, you are correct, I did suggest going to the ap_int.  forgot about that...😄.  When I suggested that I had just looked at the param itself and not the use.  Seeing how many times we cast to float supprised me.  I did a bit of reading and there is no runtime cost for a c-style cast like you have done here.  So the only advantange of going with an AP_FLOAT over an AP_INT is readability, at the cost of a few bits in eeprom.  So I think, it is completely up to you as to whichever you prefer.

bnsgeyer replied:
Thanks.  I will leave it as Int then

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -125,61 +125,81 @@ const AP_Param::GroupInfo AP_MotorsHeli_RSC::var_info[] = {
     // @Increment: 1
     // @User: Standard
-    AP_GROUPINFO("GOV_DROOP", 15, AP_MotorsHeli_RSC, _governor_droop_response, AP_MOTORS_HELI_RSC_GOVERNOR_DROOP_DEFAULT),
+    AP_GROUPINFO("GOV_DROOP", 21, AP_MotorsHeli_RSC, _governor_droop_response, 25),
```
I believe Chris did this because the new use of this parameter changed and there wasn't a good way to scale it to a new value in the upgrade.  So he just gave it a new slot to force users to re-tune when they upgrade

Gone4Dirt replied:
Yeah, this makes sense, it should be a new param key as has been done.

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -418,3 +435,81 @@ float AP_MotorsHeli_RSC::calculate_desired_throttle(float collective_in)
+    // autothrottle main power loop with governor
+    if (_governor_engage && !_governor_fault) {
+        // calculate droop - difference between actual and desired speed
+        float governor_droop = ((float)_governor_rpm - _rotor_rpm) * get_governor_droop_response();
```
In one of your previous comments (one of your first ones when you initially reviewed this PR), you asked why this couldn't be kept as a AP_Int16.  Again I believe the desire was to keep this as an AP_Int16 to reduce eeprom size?

Gone4Dirt replied:
see comment my re _cooldown_time.  Same applies here.

bnsgeyer replied:
Thanks! I'll leave as Int.

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -125,61 +125,81 @@ const AP_Param::GroupInfo AP_MotorsHeli_RSC::var_info[] = {
-    // @Range: 0 40
+    // @Param: GOV_TORQUE
+    // @DisplayName: Governor Torque Limiter
+    // @Description: Adjusts the engine's percentage of torque rise on autothrottle during ramp-up to governor speed. The torque rise will determine how fast the rotor speed will ramp up when rotor speed reaches 50% of the rotor RPM setting. The sequence of events engaging the governor is as follows: Throttle ramp time will engage the clutch and start the main rotor turning. The collective should ...
```
I believe he is referring to flat pitch or zero blade pitch angle.  Not entirely sure though

Gone4Dirt replied:
@MidwestAire could you provide some clarification on what you mean by feather pitch please?

bnsgeyer replied:
He means flat pitch or the collective pitch that is used for ground run.  I will use flat pitch in the description. Are happy with that phrase?

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -125,61 +125,81 @@ const AP_Param::GroupInfo AP_MotorsHeli_RSC::var_info[] = {
-    // @Range: 0 40
+    // @Param: GOV_TORQUE
+    // @DisplayName: Governor Torque Limiter
+    // @Description: Adjusts the engine's percentage of torque rise on autothrottle during ramp-up to governor speed. The torque rise will determine how fast the rotor speed will ramp up when rotor speed reaches 50% of the rotor RPM setting. The sequence of events engaging the governor is as follows: Throttle ramp time will engage the clutch and start the main rotor turning. The collective should ...
```
done!

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -231,65 +263,51 @@ void AP_MotorsHeli_RSC::output(RotorControlState state)
                 // set rotor control speed to idle speed parameter, this happens instantly and ignores ramping
-                _control_output = get_idle_output();
+                if (_cooldown_time > 0) {
+                    _control_output = get_idle_output() * 1.5f;
```
I don't know for sure but I trust that Chris has a better handle on this and I think it is worth waiting for a comment from a user rather than adding a parameter.  We have enough of those already.  Someday I hope someone incorporates the ICE library into the tradheli RSC and it handles this type of stuff.  So my vote is to leave it as is.

Gone4Dirt replied:
That makes sense.

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -418,3 +435,81 @@ float AP_MotorsHeli_RSC::calculate_desired_throttle(float collective_in)
+    // autothrottle main power loop with governor
+    if (_governor_engage && !_governor_fault) {
+        // calculate droop - difference between actual and desired speed
+        float governor_droop = ((float)_governor_rpm - _rotor_rpm) * get_governor_droop_response();
```
Done!

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -418,3 +435,81 @@ float AP_MotorsHeli_RSC::calculate_desired_throttle(float collective_in)
+    if (_governor_engage && !_governor_fault) {
+        // calculate droop - difference between actual and desired speed
+        float governor_droop = ((float)_governor_rpm - _rotor_rpm) * get_governor_droop_response();
+        _governor_output = governor_droop + ((throttlecurve - _governor_torque_reference) *  get_governor_ff());
```
Done!

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -418,3 +435,81 @@ float AP_MotorsHeli_RSC::calculate_desired_throttle(float collective_in)
+        _governor_output = governor_droop + ((throttlecurve - _governor_torque_reference) *  get_governor_ff());
+        if (_rotor_rpm < ((float)_governor_rpm - 2.0f)) {
+            _governor_torque_reference += get_governor_compensator();   // torque compensator
+        } else if (_rotor_rpm > ((float)_governor_rpm + 2.0f)) {
```
Done!

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -418,3 +435,81 @@ float AP_MotorsHeli_RSC::calculate_desired_throttle(float collective_in)
+        // this limits the max torque rise the governor could call for from the main power loop
+        if (_rotor_rpm >= (_governor_rpm * 0.5f)) {
+            float torque_limit = (get_governor_torque() * get_governor_torque());
+            if (_governor_rpm > 0) {
```
@Gone4Dirt This is meant as a divide by zero protection.  That's all.

Gone4Dirt replied:
My comment is still valid though.  There are a lot of nested ifs which make this code hard to read.  So this is just one easy win to reduce that, and it prevents all of this code from being executed if the governer Rpm is set to zero.

bnsgeyer replied:
got it.  done! Thanks!

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -418,3 +435,81 @@ float AP_MotorsHeli_RSC::calculate_desired_throttle(float collective_in)
+    // if governor is not engaged and rotor is overspeeding by more than 2% due to misconfigured
+    // throttle curve or stuck throttle, set a fault and governor will not operate
+    if (!_governor_engage && !_governor_fault && (_rotor_rpm > (_governor_rpm + _governor_range))) {
+        _governor_fault = true;
```
Done!

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -125,61 +125,72 @@ const AP_Param::GroupInfo AP_MotorsHeli_RSC::var_info[] = {
+    // @Description: Feedforward governor gain to throttle response during sudden loading/unloading of the rotor system. If Rrpm drops excessively during full collective climb with the droop response set correctly, increase the governor feedforward. If Rrpm drops excessively under heavy load also inspect the setting for maximum collective pitch to ensure that the pitch setting corresponds to maxi ...
+    // @Range: 0 100
     // @Units: %
     // @Increment: 1
```
Done!

bnsgeyer commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -125,61 +125,81 @@ const AP_Param::GroupInfo AP_MotorsHeli_RSC::var_info[] = {
+
+    // @Param: CLDWN_TIME
+    // @DisplayName: Cooldown Time
+    // @Description: Will provide a fast idle for engine cooldown by raising the Ground Idle speed setting by 50% for the number of seconds the timer is set for. A setting of zero disables the fast idle. This feature will only apply after the runup complete has been declared. At any time during fast idle, disarming will shut the engine down however the timer to declare below critical speed (which ...
```
I have changed the behavior so this statement doesn’t apply. The cool down timer doesn’t affect the autonomous shutdown.

Gone4Dirt reviewed (approved):
LGTM
